FNML/ontology.md

```diff
@@ -92,21 +109,81 @@ see, e.g. test case [RMLFNOTC0019](https://github.com/RMLio/rml-fno-test-cases/t
 The mapping challenge [Join on literals](https://github.com/kg-construct/mapping-challenges/pull/29) also influences this spec.
 </p>
 
-#### Subject map
+### fnml:execution
 
-When this triples map **does not specify a subject map**, a blank node should be generated for the subject.
+fnml:execution connects the RDF dataset generating triples map via a [fnml:ExecutionTermMap] with a [=fnml:Execution=].
+It has domain [=fnml:ExecutionTermMap=] and range [=fnml:Execution=].
 
-<div class="note">
-We should probably define a new resource that is a subclass of rr:TriplesMap,
-as the description above is disjoint with rr:TriplesMap, as logical source and subjectMap are not optional in [[R2RML]]
-</div>
+### fnml:output
+
+<dfn class="lint-ignore">fnml:output</dfn> connects the RDF dataset generating triples map via a [fnml:ExecutionTermMap] with an output predicate.
+It has domain [=fnml:ExecutionTermMap=] and range [rdf:Property].
+
+### fnml:function
+
+<dfn class="lint-ignore">fnml:function</dfn> connects the [fnml:Execution] with an FnO [=function description=].
+It has domain [=fnml:Execution=] and range [fno:Function](https://w3id.org/function/ontology#Function).
+
+### fnml:predicate
+
+<dfn class="lint-ignore">fnml:predicate</dfn> connects the [fnml:ParameterTermMap] with a parameter predicate.
+It has domain [=fnml:ParameterTermMap=] and range [rdf:Property].
 
 ### Nested functions
 
-As the range of `fnml:functionValue` is `rr:TriplesMap`,
+As [=fnml:ParameterTermMap=] is a subclass of [=term map=],
 it is possible to nest functions: you generate a term in a first function, and that term is used as an parameter value in a second function.
-For an example, see [RMLFNOTC0018](https://github.com/RMLio/rml-fno-test-cases/tree/master/RMLFNOTC0018-CSV).
+
+<p class="issue" data-format="markdown">
+For an old example, see [RMLFNOTC0018](https://github.com/RMLio/rml-fno-test-cases/tree/master/RMLFNOTC0018-CSV).
+</p>
 
 <p class="issue" data-number="3" data-format="markdown">
 For now, it is unclear how to handle a nested function where that nested triplesmap contains a join condition.
 </p>
+
+```turtle "example": "using toUppercase in an RML mapping"
+<#Person_Mapping>
+    rml:logicalSource <#LogicalSource> ;
+    rr:subjectMap <#SubjectMap> ;
+    rr:predicateObjectMap <#NameMapping> .
+
+<#NameMapping>
+    rr:predicate dbo:title ;
+    rr:objectMap <#ExecutionTermMap> .
+
+<#ExecutionTermMap> a fnml:ExecutionTermMap ;
+    fnml:execution <#Execution> ;
+    fnml:output grel:stringOutput .
+
+<#Execution> a fnml:Execution ;
+    fnml:function grel:toUppercase ;
+    fnml:inputParameters
+        [
+            a fnml:ParameterTermMap
+            fnml:predicate grel:stringInput ;
+            fnml:execution <#Execution2> ;     # Link to the nested execution
+            fnml:output grel:stringOutput .    # Specify which output of the nested execution is taken
+        ] .
```

Comment:
Hmm this is tricky. A ParameterTermMap is now not only a `TermMap` and `PredicateObjectMap`, but also sometimes an `ExecutionTermMap`?

Not saying it can't work, but, I wonder if it wouldn't be cleaner to keep these concepts seperate?

e.g.

```
        [
            a fnml:ParameterMap
            fnml:parameter grel:stringInput ;
            fnml:parameterValue [
                # a rr:TermMap, fnml:ExecutionTermMap
                fnml:execution <#Execution2> ;     # Link to the nested execution
                fnml:output grel:stringOutput .    # Specify which output of the nested execution is taken
            ]
        ] .
```

In this case I introduce a new predicate `fnml:parameterValue` which refers to a `TermMap`, thus making it natural to reuse `fnml:ExecutionTermMap` at this location.